src/sharedData/visualization/components/VisualizationMapLayer.js

```diff
@@ -0,0 +1,381 @@
+        ? null
+        : new mapboxgl.Popup({
+            className: `${sourceName}-marker-popup`,
+          }).setDOMContent(popupContainer),
+    [popupContainer, isMapFile, sourceName]
+  );
+
+  useEffect(() => {
+    if (!visualizationConfig?.visualizeConfig) {
+      return;
+    }
+    getLayerImage({
+      ...(visualizationConfig?.visualizeConfig || {}),
+      opacity: DEFAULT_MARKER_OPACITY,
+    }).then(image => {
+      return setimageSvg(image);
+    });
+  }, [visualizationConfig?.visualizeConfig]);
+
+  const openPopup = useCallback(
+    (feature, event) => {
+      if (!feature || isMapFile) {
+        return;
+      }
+      const coordinates = feature.geometry.coordinates;
+
+      if (event) {
+        // Ensure that if the map is zoomed out such that
+        // multiple copies of the feature are visible, the
+        // popup appears over the copy being pointed to.
+        while (Math.abs(event.lngLat.lng - coordinates[0]) > 180) {
+          coordinates[0] += event.lngLat.lng > coordinates[0] ? 360 : -360;
+        }
+      }
+
+      setPopupData({
+        coordinates,
+        data: feature.properties,
+      });
+    },
+    [isMapFile]
+  );
+
+  useEffect(() => {
+    if (!map || !popupData?.coordinates || isMapFile) {
+      return;
+    }
+    popup.setLngLat(popupData?.coordinates);
+    if (!popup.isOpen()) {
+      popup.addTo(map);
+    }
+  }, [popup, popupData?.coordinates, map, isMapFile]);
+
+  useEffect(() => {
+    if (!showPopup || !map) {
+      return;
+    }
+    const source = map.getSource(sourceName);
+    if (!source) {
+      return;
+    }
+    const features = source?._data?.features;
+    if (_.isEmpty(features)) {
+      return;
+    }
+    openPopup(features[0]);
+  }, [
+    showPopup,
+    openPopup,
+    map,
+    visualizationConfig?.annotateConfig?.annotationTitle,
+    visualizationConfig?.annotateConfig?.dataPoints,
+    sourceName,
+  ]);
+
+  useEffect(() => {
+    if (!map) {
+      return;
+    }
+    const visualizationConfigBounds = (async function () {
+      const viewportBounds = visualizationConfig?.viewportConfig?.bounds;
+      if (!viewportBounds) {
+        return;
+      }
+      const southWest = viewportBounds.southWest;
+      const northEast = viewportBounds.northEast;
+      if (!southWest || !northEast) {
+        return;
+      }
+      const sw = new mapboxgl.LngLat(southWest.lng, southWest.lat);
+      const ne = new mapboxgl.LngLat(northEast.lng, northEast.lat);
+      return new mapboxgl.LngLatBounds(sw, ne);
+    })();
+
+    const sourceBounds = getSourceBounds(map, sourceName);
+
+    Promise.all([visualizationConfigBounds, sourceBounds]).then(
+      ([visualizationConfigBounds, sourceBounds]) => {
+        if (!changeBounds) {
+          return;
+        }
+        const bounds =
+          useConfigBounds && visualizationConfigBounds
+            ? visualizationConfigBounds
+            : sourceBounds;
+
+        if (bounds && !bounds.isEmpty()) {
```

Comment:
question: what does it mean when you have a bounds but it's empty?

Reply:
@paulschreiber empty means that it has no values, this is a Mapbox data structure, from the ducmentation:
```
Check if the bounding box is an empty/null-type box.
```